pkg/plugin/plugin.go

```diff
@@ -0,0 +1,286 @@
+package plugin
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+type LocalVolumeObjectStore struct {
+	log        logrus.FieldLogger
+	volumeType VolumeType
+}
+
+// NewFileObjectStore instantiates a LocalVolumeObjectStore.
+func NewLocalVolumeObjectStore(log logrus.FieldLogger, v VolumeType) *LocalVolumeObjectStore {
+	return &LocalVolumeObjectStore{
+		log:        log,
+		volumeType: v,
+	}
+}
+
+// Init initializes the plugin. It can be called multiple times.
+func (o *LocalVolumeObjectStore) Init(config map[string]string) error {
+	bucket := config["bucket"]
+	path := filepath.Join(getRoot(), bucket)
+
+	log := o.log.WithFields(logrus.Fields{
+		"bucket": bucket,
+		"path":   path,
+	})
+	log.Debug("LocalPvObjectStore.Init called")
+
+	// Get the bucket, check if it exists
+	info, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		log.Info("Bucket/Volume does not already exist. Initializing.")
+
+		var opts localVolumeObjectStoreOpts
+		pluginConfigMap, err := getPluginConfig(o.volumeType)
+		if err != nil {
+			return errors.Wrap(err, "failed to get plugin config map")
+		}
+		if pluginConfigMap == nil {
+			log.Debug("Did not find a configmap fot this plugin")
+			opts = localVolumeObjectStoreOpts{}
+		} else {
+			log.Debug("Found a configmap for this plugin")
+			opts = localVolumeObjectStoreOpts{
+				veleroDeploymentName: pluginConfigMap.Data["veleroDeploymentName"],
+				resticDaemonsetName:  pluginConfigMap.Data["resticDaemonsetName"],
+				fileserverImage:      pluginConfigMap.Data["fileserverImage"],
+			}
+		}
+
+		deployment, err := getDeployment(opts)
+		if err != nil {
+			return errors.Wrap(err, "could not get Velero deployment")
+		}
+
+		ds, err := peakDameonset(opts)
+		if err != nil {
+			return errors.Wrap(err, "could not get restic daemonset")
+		}
+
+		volumeSpec, err := buildVolume(o.volumeType, config)
+		if err != nil {
+			return errors.Wrap(err, "failed to build volume")
+		}
+
+		volumeMountSpec := buildVolumeMount(bucket, path)
+
+		// If restic is present, it must also mount the volume
+		if ds != nil {
+			err = ensureDaemonsetHasVolume(ds, volumeSpec, volumeMountSpec)
+			if err != nil {
+				return errors.Wrap(err, "failed to update restic daemonset")
+			}
+		}
+
+		err = ensureDeploymentHasVolume(deployment, volumeSpec, volumeMountSpec, opts)
+		if err != nil {
+			return errors.Wrap(err, "failed to update velero deployment")
+		}
+
+		return errors.New("volume initialized, restart pending")
```

Comment:
any way to force the restart here?

Reply:
No, since the plugin runs as a separate process, even calling `os.Exit()` won't really restart the pod.

Reply:
What about getting the Velero k8s deployment and deleting it?  Not sure that's idiomatic to do from inside of the deployment being deleted, feel free to close this one out if that's a terrible idea

Reply:
Not sure I want to delete the deployments since I just updated it. I could try to delete the pod, but I'm not sure I want to thrash the controller by maybe having it try to restart the stale pod config before the deployment controller picks up the update.